Source/Core/VideoBackends/Metal/MTLObjectCache.mm

```diff
@@ -0,0 +1,512 @@
+  }
+}
+
+static MTLSamplerAddressMode Convert(WrapMode wrap)
+{
+  switch (wrap)
+  {
+  case WrapMode::Clamp:  return MTLSamplerAddressModeClampToEdge;
+  case WrapMode::Mirror: return MTLSamplerAddressModeMirrorRepeat;
+  case WrapMode::Repeat: return MTLSamplerAddressModeRepeat;
+  }
+}
+
+static const char* to_string(FilterMode filter)
+{
+  switch (filter)
+  {
+  case FilterMode::Linear: return "Ln";
+  case FilterMode::Near:   return "Pt";
+  }
+}
+static const char* to_string(WrapMode wrap)
+{
+  switch (wrap)
+  {
+  case WrapMode::Clamp:  return "C";
+  case WrapMode::Mirror: return "M";
+  case WrapMode::Repeat: return "R";
+  }
+}
+
+// clang-format on
+
+MRCOwned<id<MTLSamplerState>> Metal::ObjectCache::CreateSampler(SamplerSelector sel)
+{
+  @autoreleasepool
+  {
+    auto desc = MRCTransfer([MTLSamplerDescriptor new]);
+    [desc setMinFilter:ConvertMinMag(sel.MinFilter())];
+    [desc setMagFilter:ConvertMinMag(sel.MagFilter())];
+    [desc setMipFilter:ConvertMip(sel.MipFilter())];
+    [desc setSAddressMode:Convert(sel.WrapU())];
+    [desc setTAddressMode:Convert(sel.WrapV())];
+    [desc setMaxAnisotropy:1 << (sel.AnisotropicFiltering() ? g_ActiveConfig.iMaxAnisotropy : 0)];
+    [desc setLabel:MRCTransfer([[NSString alloc]
+                       initWithFormat:@"%s%s%s %s%s%s", to_string(sel.MinFilter()),
+                                      to_string(sel.MagFilter()), to_string(sel.MipFilter()),
+                                      to_string(sel.WrapU()), to_string(sel.WrapV()),
+                                      sel.AnisotropicFiltering() ? "(AF)" : ""])];
+    return MRCTransfer([Metal::g_device newSamplerStateWithDescriptor:desc]);
+  }
+}
+
+void Metal::ObjectCache::ReloadSamplers()
+{
+  for (auto& sampler : m_samplers)
+    sampler = nullptr;
+}
+
+// MARK: Pipelines
+
+static MTLPrimitiveTopologyClass GetClass(PrimitiveType prim)
+{
+  switch (prim)
+  {
+  case PrimitiveType::Points:
+    return MTLPrimitiveTopologyClassPoint;
+  case PrimitiveType::Lines:
+    return MTLPrimitiveTopologyClassLine;
+  case PrimitiveType::Triangles:
+  case PrimitiveType::TriangleStrip:
+    return MTLPrimitiveTopologyClassTriangle;
+  }
+}
+
+static MTLPrimitiveType Convert(PrimitiveType prim)
+{
+  // clang-format off
+  switch (prim)
+  {
+  case PrimitiveType::Points:        return MTLPrimitiveTypePoint;
+  case PrimitiveType::Lines:         return MTLPrimitiveTypeLine;
+  case PrimitiveType::Triangles:     return MTLPrimitiveTypeTriangle;
+  case PrimitiveType::TriangleStrip: return MTLPrimitiveTypeTriangleStrip;
+  }
+  // clang-format on
+}
+
+static MTLCullMode Convert(CullMode cull)
+{
+  switch (cull)
+  {
+  case CullMode::None:
+  case CullMode::All:  // Handled by disabling rasterization
+    return MTLCullModeNone;
+  case CullMode::Front:
+    return MTLCullModeFront;
+  case CullMode::Back:
+    return MTLCullModeBack;
+  }
+}
+
+static MTLBlendFactor Convert(DstBlendFactor factor, bool src1)
```

Comment:
This feels like something that should be changed in `BlendingState`, since this is just a forward of `BlendingState.usedualsrc`

Maybe I should name it `usedualsrc` though, that might make sense